File: matlab/src/lang/processes/MMDP2.m
classdef MMDP2 < MMDP
    % 2-state Markov-Modulated Deterministic Process
    %
    % A specialized MMDP with exactly 2 phases, using a convenient
    % parameterization analogous to MMPP2.
    %
    % Parameterization:
    %   r0, r1: Deterministic rates in states 0 and 1
    %   sigma0: Transition rate from state 0 to state 1
    %   sigma1: Transition rate from state 1 to state 0
    %
    % The generator matrix is:
    %   Q = [-sigma0, sigma0; sigma1, -sigma1]
    %
    % The rate matrix is:
    %   R = diag([r0, r1])
    %
    % Copyright (c) 2012-2026, Imperial College London
    % All rights reserved.

    methods
        function self = MMDP2(r0, r1, sigma0, sigma1)
            % MMDP2 Create a 2-state Markov-Modulated Deterministic Process
            %
            % @brief Creates a 2-state MMDP with specified rates and transitions
            % @param r0 Deterministic rate in state 0
            % @param r1 Deterministic rate in state 1
            % @param sigma0 Transition rate from state 0 to state 1
            % @param sigma1 Transition rate from state 1 to state 0
            % @return self MMDP2 instance

            Q = [-sigma0, sigma0; sigma1, -sigma1];
            R = diag([r0; r1]);
            self@MMDP(Q, R);
            self.name = 'MMDP2';

            % Override parameters with original scalar values for convenience
            self.params = cell(1, 4);
            setParam(self, 1, 'r0', r0);
            setParam(self, 2, 'r1', r1);
            setParam(self, 3, 'sigma0', sigma0);
            setParam(self, 4, 'sigma1', sigma1);
        end

        function rate = getMeanRate(self)
            % GETMEANRATE Compute stationary mean rate (closed-form)
            %
            % For a 2-state MMDP, the mean rate has the closed form:
            %   E[r] = (r0*σ1 + r1*σ0) / (σ0 + σ1)
            %
            % @return rate Stationary mean deterministic rate

            r0 = self.getParam(1).paramValue;
            r1 = self.getParam(2).paramValue;
            sigma0 = self.getParam(3).paramValue;
            sigma1 = self.getParam(4).paramValue;
            rate = (r0*sigma1 + r1*sigma0) / (sigma0 + sigma1);
        end

        function scv = getSCV(self)
            % GETSCV Compute squared coefficient of variation (closed-form)
            %
            % For a 2-state MMDP, the SCV has a closed form based on
            % the variance of rates over the stationary distribution.
            %
            % @return scv Squared coefficient of variation

            r0 = self.getParam(1).paramValue;
            r1 = self.getParam(2).paramValue;
            sigma0 = self.getParam(3).paramValue;
            sigma1 = self.getParam(4).paramValue;

            % Stationary probabilities
            pi0 = sigma1 / (sigma0 + sigma1);
            pi1 = sigma0 / (sigma0 + sigma1);

            % Mean and variance
            mean_rate = pi0 * r0 + pi1 * r1;
            var_rate = pi0 * r0^2 + pi1 * r1^2 - mean_rate^2;

            if mean_rate > 0
                scv = var_rate / (mean_rate^2);
            else
                scv = Inf;
            end
        end

        function n = getNumberOfPhases(~)
            % GETNUMBEROFPHASES Return the number of phases
            %
            % @return n Always 2 for MMDP2

            n = 2;
        end

        function Q_mat = Q(self)
            % Q Return the generator matrix
            %
            % @return Q_mat 2×2 generator matrix

            sigma0 = self.getParam(3).paramValue;
            sigma1 = self.getParam(4).paramValue;
            Q_mat = [-sigma0, sigma0; sigma1, -sigma1];
        end

        function R_mat = R(self)
            % R Return the rate matrix (diagonal)
            %
            % @return R_mat 2×2 diagonal rate matrix

            r0 = self.getParam(1).paramValue;
            r1 = self.getParam(2).paramValue;
            R_mat = diag([r0; r1]);
        end

        function r_vec = r(self)
            % r Return the rate vector
            %
            % @return r_vec 2-vector of rates [r0; r1]

            r0 = self.getParam(1).paramValue;
            r1 = self.getParam(2).paramValue;
            r_vec = [r0; r1];
        end
    end
end

File: jar/src/main/java/jline/lang/processes/MMDP2.java
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */

package jline.lang.processes;

import jline.util.matrix.Matrix;
import jline.util.matrix.MatrixCell;

import java.io.Serializable;

/**
 * A 2-state Markov-Modulated Deterministic Process.
 *
 * Specialized MMDP with exactly 2 phases, using a convenient
 * parameterization analogous to MMPP2.
 *
 * Parameterization:
 * <ul>
 *   <li>r0, r1: Deterministic rates in states 0 and 1</li>
 *   <li>sigma0: Transition rate from state 0 to state 1</li>
 *   <li>sigma1: Transition rate from state 1 to state 0</li>
 * </ul>
 *
 * The generator matrix is:
 *   Q = [-sigma0, sigma0; sigma1, -sigma1]
 *
 * The rate matrix is:
 *   R = diag([r0, r1])
 */
public class MMDP2 extends MMDP implements Serializable {

    private double r0;
    private double r1;
    private double sigma0;
    private double sigma1;

    /**
     * Creates a 2-state Markov-Modulated Deterministic Process.
     *
     * @param r0     deterministic rate in state 0
     * @param r1     deterministic rate in state 1
     * @param sigma0 transition rate from state 0 to state 1
     * @param sigma1 transition rate from state 1 to state 0
     */
    public MMDP2(double r0, double r1, double sigma0, double sigma1) {
        super(buildQ(sigma0, sigma1), buildR(r0, r1));
        this.name = "MMDP2";
        this.r0 = r0;
        this.r1 = r1;
        this.sigma0 = sigma0;
        this.sigma1 = sigma1;

        // Override parameters with original scalar values for convenience
        this.setParam(1, "r0", r0);
        this.setParam(2, "r1", r1);
        this.setParam(3, "sigma0", sigma0);
        this.setParam(4, "sigma1", sigma1);
    }

    private static Matrix buildQ(double sigma0, double sigma1) {
        Matrix Q = new Matrix(2, 2);
        Q.set(0, 0, -sigma0);
        Q.set(0, 1, sigma0);
        Q.set(1, 0, sigma1);
        Q.set(1, 1, -sigma1);
        return Q;
    }

    private static Matrix buildR(double r0, double r1) {
        Matrix R = new Matrix(2, 2);
        R.set(0, 0, r0);
        R.set(1, 1, r1);
        return R;
    }

    /**
     * Returns the generator matrix Q (closed-form for 2 states).
     *
     * @return 2x2 generator matrix
     */
    @Override
    public Matrix Q() {
        return buildQ(sigma0, sigma1);
    }

    /**
     * Returns the rate matrix R (diagonal, closed-form for 2 states).
     *
     * @return 2x2 diagonal rate matrix
     */
    @Override
    public Matrix R() {
        return buildR(r0, r1);
    }

    /**
     * Returns the rate vector (diagonal of R).
     *
     * @return 2-vector of rates [r0; r1]
     */
    @Override
    public Matrix r() {
        Matrix r = new Matrix(2, 1);
        r.set(0, 0, r0);
        r.set(1, 0, r1);
        return r;
    }

    @Override
    public long getNumberOfPhases() {
        return 2;
    }

    /**
     * Computes the stationary mean rate (closed-form).
     *
     * For a 2-state MMDP, the mean rate has the closed form:
     *   E[r] = (r0*sigma1 + r1*sigma0) / (sigma0 + sigma1)
     *
     * @return stationary mean deterministic rate
     */
    @Override
    public double getMeanRate() {
        return (r0 * sigma1 + r1 * sigma0) / (sigma0 + sigma1);
    }

    /**
     * Computes the squared coefficient of variation (closed-form).
     *
     * For a 2-state MMDP, the SCV has a closed form based on
     * the variance of rates over the stationary distribution.
     *
     * @return squared coefficient of variation
     */
    @Override
    public double getSCV() {
        // Stationary probabilities
        double pi0 = sigma1 / (sigma0 + sigma1);
        double pi1 = sigma0 / (sigma0 + sigma1);

        // Mean and variance
        double meanRate = pi0 * r0 + pi1 * r1;
        double varRate = pi0 * r0 * r0 + pi1 * r1 * r1 - meanRate * meanRate;

        if (meanRate > 0) {
            return varRate / (meanRate * meanRate);
        }
        return Double.POSITIVE_INFINITY;
    }

    @Override
    public String toString() {
        return String.format("MMDP2(r0=%.6f, r1=%.6f, sigma0=%.6f, sigma1=%.6f)",
                r0, r1, sigma0, sigma1);
    }
}
